### ISSUE TITLE
Better tabs detection

### ISSUE THREAD

**Issue by lwouis** [lwouis/alt-tab-macos#1540]
Today, AltTab uses a full bag of tricks to detect windows which are grouped as tabs. These tricks have limitations that result in features we can't provide, and bugs where AltTab fails to show or not show certain windows.

### Current implementation (very simplified and high-level)

* AltTab subscribe to the Accessibility event called `kAXFocusedUIElementChangedNotification` on every app
* The event triggers when the user interacts with the app UI in almost any way
* On every trigger, we make an API call to ask for the app windows
* If the windows returned by the API are different from some of the windows we already were tracking, it means some windows just became tabs

### Limitations / Bugs

* If there are multiple tab-groups for a given app, AltTab doesn't know which tab-group has which tabs. All it knows is which windows are the active tabs of the tab-groups.
* Tab detection seems to be flacky, especially when dealing with animations like Space transitions, windows going full-screen, etc. Overall, the `kAXFocusedUIElementChangedNotification` doesn't cover all use-cases for tabs, and doesn't happen always at the right time (i.e. it may trigger too early or too late, and we get the wrong understanding of the OS state)

**Comment by lwouis** [comment]
FYI, we can't use the Accessibility API to simply ask the tab-group for its children. The API simply doesn't return children windows. We can only ask children windows of the whole app, and we will get a single list, with no sorting depending on which tab-group the windows are in.

Furthermore, the OS handles tabs in a lazy way. A window becoming a tab will not be resized, repositioned, and redrawned, until it is clicked/seen by the user. So after a user makes a window a tab, if they don't click on it, and we make an Accessibility call to get its size, position, or screenshot, we will get the values of when it used to be a window - so completely wrong.

**Comment by lwouis** [comment]
I looked into the following APIs:

```swift
SLSCopyAssociatedWindows
```

It returns the same ID as the one given as input

```swift
let query = SLSWindowQueryWindows(cgsMainConnectionId, [try currentWindows![0].cgWindowId()!] as CFArray, 1)
let iterator = SLSWindowQueryResultCopyWindows(query)
while SLSWindowIteratorAdvance(iterator) == .success {
  let parent_wid = SLSWindowIteratorGetParentID(iterator)
  let wid = SLSWindowIteratorGetWindowID(iterator)
  let tags = SLSWindowIteratorGetTags(iterator)
}
```

`SLSWindowIteratorGetParentID` returns `0` (window is top level / has no parent, I guess)
`SLSWindowIteratorGetWindowID` returns the same ID as input (window has no child, I guess)
`SLSWindowIteratorGetTags` returns a bitmask which I don't know how to break down and what it represents

**Comment by lwouis** [comment]
- Using `CGSCopyWindowsWithOptionsAndTags` doesn't show tabbed windows
- Using `CGWindowListCopyWindowInfo` shows tabbed windows
  - Their `kCGWindowBounds` are not up-to-date (see my comment above about tabs being updated lazily by macOS)
  - Their `kCGWindowAlpha`, `kCGWindowLayer` are not helpful to discriminate tabs
  - Their `kCGWindowIsOnscreen` returns `nil` for tabbed windows, and `true` for non-tabbed windows. However, it may also return `nil` for other reasons than tabs (e.g. window is minimized), so probably unreliable
- On the `AXref` of a window which has tabs, there is nothing we can get except the names of the tabs (not enough to ID the tab)  and their count. I tried `kAXIdentifierAttribute`, `kAXWindowAttribute`, `kAXWindowsAttribute`, `kAXMainWindowAttribute`, `kAXFocusedWindowAttribute`, `kAXEnabledAttribute`, `kAXContentsAttribute`, `kAXTabsAttribute`, but nothing is helpful here.

So at a given point of time, without prior knowledge, there is no way to know which windows are tabbed. That's why today we look for the only event that triggers when the user tabs/un-tabs a window: `kAXFocusedUIElementChangedNotification`. It's really unfortunate it's the only event we can observe to know when to check if a window disappeared (which we assume is a tab then) because that event creates issue in many apps (e.g. IntelliJ, Matlab, LibreOffice, Android Studio, etc) and even in macOS itself (#718).

**Comment by yossizahn** [comment]
Since the latest release Alt-tab has started recognizing tabs in [Fork](https://git-fork.com/) and [Tableplus](https://tableplus.com/) as seperate windows. 
Safari and Iterm2 tabs don't show as seperate windows.

**Comment by lwouis** [comment]
@yossizahn Safari and iTerm2 tabs aren't standard tabs. They are custom UI painted inside an NSWindow. So there is no chance AltTab can get information about them.

Imagine that it's only when you click on a non-active tab that these apps will draw their content on screen. We can't get the screenshot of these since they are not NSWindow instance and are rendered lazily when shown, on top of the same NSWindow instance.

That's why the preference is called "Show standard tabs as separate windows" and not "Show tabs as separate windows". Standard tabs are specific tech from macOS SDK that lets you merge windows together seamlessly. What these 2 apps are doing is custom UI / custom rendering of pixels on screen.

**Comment by yossizahn** [comment]
@lwouis Thanks for the reply, but I still don't understand.
I have always had "Show standard tabs as separate windows" switched off, and I was never shown tabs as seperate windows. In a recent release something changed and I am show tabs as seperate windows regardless of the setting. 
Is this a bug or am I missing somthing?
Thanks

**Comment by lwouis** [comment]
Yes, sorry @yossizahn, I forgot to update you here when I realized through #1656 that it was an actual regression. Hopefully you can use the latest version and your problem will be gone by then~

**Comment by yossizahn** [comment]
No need to apologize, thanks for a great app!

**Comment by lwouis** [comment]
For posteriti, I looked into old APIs to check if we could detect tabs with them. Still not possible today.

```swift
// SLSCopyAssociatedWindows
let wins = SLSCopyAssociatedWindows(CGS_CONNECTION, wid)
Logger.error { wins as! [CGWindowID] }

// SLSWindowQueryWindows
let query = SLSWindowQueryWindows(CGS_CONNECTION, [wid] as CFArray, 1)
let iterator = SLSWindowQueryResultCopyWindows(query)
if (SLSWindowIteratorGetCount(iterator) == 1) {
    if (SLSWindowIteratorAdvance(iterator)) {
        let parent_wid = SLSWindowIteratorGetParentID(iterator)
        Logger.error { parent_wid }
    }
}

// SLSCopyWindowGroup
var windowList = [] as CFArray
var windowCount = 10 as UInt
SLSCopyWindowGroup(CGS_CONNECTION, wid, "movementGroup" as CFString, &windowList, &windowCount)
Logger.error { (windowList as! [CGWindowID]) }
```

What is prints for 2 Finder windows that are tabbed together:

```
Window: (pid:817 com.apple.finder) (wid:2175 title:Movies)
[2175]
0
[]
Window: (pid:817 com.apple.finder) (wid:2173 title:Desktop)
[2173]
0
[]
```

Pretty useless. It could be used to detect sheets at least. This is what it prints for a TextEdit window with a Print sheet on it:

```
Window: (pid:803 com.apple.TextEdit) (wid:73 title:tmp.rtf)
[2292, 73]
0
[]
```
